File: elixir/lib/2020/01_report_repair/aoc202001.ex
defmodule AOC2020.Day01 do
  @moduledoc """
  Advent of Code 2020, day 1: Report Repair
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n")
    |> Enum.map(&String.to_integer/1)
    |> MapSet.new()
  end

  @doc """
  Solve part 1
  """
  def part1(input) do
    input |> find_summands() |> Enum.product()
  end

  @doc """
  Solve part 2
  """
  def part2(input) do
    input
    |> Enum.find_value(fn first ->
      case find_summands(input, 2020 - first) do
        [0, _] -> nil
        [second, third] -> [first, second, third]
      end
    end)
    |> Enum.product()
  end

  @doc """
  Find the two numbers in the list that sum to target

  ## Examples:

      iex> find_summands([1721, 979, 299])
      [1721, 299]

      iex> find_summands([1721, 979, 299], 2021)
      [0, 2021]
  """
  def find_summands(numbers, target \\ 2020) do
    numbers
    |> Enum.find(0, fn first -> (target - first) in numbers end)
    |> then(fn first -> [first, target - first] end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: julia/2020/01_report_repair/aoc202001.jl
# Advent of Code 2020, day 1: Report Repair

module AOC202001

# Parse input
function parse_data(puzzle_input)
    [parse(Int32, n) for n ∈ split(puzzle_input, "\n")] |> Set
end


# Solve part 1
function part1(data)
    data |> find_summands |> prod
end

# Solve part 2
function part2(data)
    summands = nothing
    for first in data
        pairs = find_summands(data, 2020 - first)
        if !isempty(pairs)
            summands = [[first]; pairs]
            break
        end
    end
    summands |> prod
end

function find_summands(numbers, target = 2020)
    [n for n in numbers if target - n ∈ numbers]
end

# Solve the puzzle for the given input
function solve(puzzle_input)
    data = puzzle_input |> parse_data
    part1(data), part2(data)
end


# Solve the problem for one file
function solve_file(file_path)
    "\n$(file_path)" |> println
    open(file_path) do fid
        read(fid, String) |> strip .|> solve
    end
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
